crates/ra_hir_ty/src/infer/expr.rs

```diff
@@ -134,7 +137,12 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 // infer the body.
                 self.coerce(&closure_ty, &expected.ty);
 
-                self.infer_expr(*body, &Expectation::has_type(ret_ty));
+                self.return_tys.push(ret_ty.clone());
+
+                self.infer_expr_coerce(*body, &Expectation::has_type(ret_ty));
+
+                self.return_tys.pop();
```

Comment:
Hm, do we actually need a stack of return types? We already have a call stack!

```suggestion
let prev_ret_ty = mem::replace(&mut self.ret_ty, ret_ty.clone())
self.infer_expr_coerce(*body, &Expectation::has_type(ret_ty));
self.ret_ty = prev_ret_ty;
```

Reply:
oh, yeah, of course